ethpm/backends/registry.py

```diff
@@ -0,0 +1,39 @@
+from web3 import Web3
+
+from ethpm.backends.base import BaseURIBackend
+from ethpm.backends.ipfs import get_ipfs_backend
+from ethpm.exceptions import UriNotSupportedError
+from ethpm.utils.registry import lookup_manifest_uri_located_at_registry_uri
+from ethpm.validation import is_valid_registry_uri
+
+
+class RegistryURIBackend(BaseURIBackend):
+    """
+    Backend class to handle Registry URIs.
+
+    A Registry URI must point towards a valid IPFS URI.
+
+    `RegistryURIBackend` uses IPFS backend set via env variable
+    (defaults to InfuraIPFSBackend).
+    """
+
+    def can_handle_uri(self, uri: str) -> bool:
+        return is_valid_registry_uri(uri)
+
+    def fetch_uri_contents(self, uri: str, w3: Web3) -> bytes:
+        """
+        Return the contents stored at `uri` assuming `uri`
+        is a valid Registry URI pointing towards an IPFS URI.
+        Requires a valid web3 instance connected to the chain on which
+        the registry lives.
+        """
+        manifest_uri = lookup_manifest_uri_located_at_registry_uri(uri, w3)
```

Comment:
Is this the only place where the `w3` instance is actually used?  It'd be nice to get it out of the signature for `fetch_uri_contents` and have it be provided via some other mechanism.  Maybe as a constructor argument to this backend?

Reply:
> Also, it's possible that RegistryURIBackend.fetch_uri_contents simply returns the ipfs uri stored on the registry and then we handle fetching the ipfs uri's contents outside of this backend -- but I like the idea of this method actually fetching and returning the stored ipfs uri's contents and encapsulating all of that logic here.

I think this concept exists within the populus code which dealt with packages.  There was a mechanism that allowed backends to *resolve* a package identifier, but resolution had a *finality* concept which indicated whether the resource it was returning was fully resolved, or whether it needed to be passed through another round of backend resolution.

Reply:
Here it is:

https://github.com/ethereum/populus/blob/feat/v2/populus/utils/packaging.py#L442